### PR TITLE
Add support for binary values in TXT records in service files

### PR DESCRIPTION
**Pull request by krnowak**
This adds an attribute to the txt-record tag that describes the format
of the value part in the key=value part. For backwards-compatibility
it defaults to text if not specified. Other options include decoding
the value from hex encoding or from base64 encoding. That way we can
specify binary data in an XML file, that normally does not allow
arbitrary bytes in its contents.

These encodings result in larger data than the resulting binary data,
but that should not be a problem - TXT records ought to be small
anyway.

We need that for https://github.com/ostreedev/ostree/pull/782 where we want to send a small bloom filter as the txt record. Encoding that with base64 unnecessarily shortens the filter by 25%.

This is the alternative proposal to https://github.com/lathiat/avahi/pull/84 - the differences are:
- we don't introduce new element for binary records, but instead add an optional attribute to the txt-record element
- we don't use sscanf for reading the hex values - it will happily take "3g" and return the value 3 without saying that "g" was skipped
- we also add support for base64 encoding

### CODE REVIEW

File: avahi-daemon/static-services.c [hunk 1]
```diff
@@ -308,12 +308,21 @@ typedef enum {
...
+    TXT_RECORD_VALUE_BINARY_BASE64
+} txt_record_value_type;
+
 struct xml_userdata {
     StaticServiceGroup *group;
     StaticService *service;
     xml_tag_name current_tag;
     int failed;
     char *buf;
+    int txt_value_length;
```
Comment (pwithnall) [inline avahi-daemon/static-services.c]: I’d make that a `size_t`.

File: avahi-daemon/static-services.c [hunk 2]
```diff
@@ -308,12 +308,21 @@ typedef enum {
     XML_TAG_TXT_RECORD
 } xml_tag_name;
 
+typedef enum {
+    TXT_RECORD_VALUE_TEXT,
+    TXT_RECORD_VALUE_BINARY_HEX,
+    TXT_RECORD_VALUE_BINARY_BASE64
```
Comment (pwithnall) [inline avahi-daemon/static-services.c]: Nitpick: Missing comma on the last element (reduces diff noise if further elements are added in future).

File: avahi-daemon/static-services.c [hunk 3]
```diff
@@ -403,8 +412,29 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
 
         u->current_tag = XML_TAG_PORT;
     } else if (u->current_tag == XML_TAG_SERVICE && strcmp(el, "txt-record") == 0) {
-        if (attr[0])
-            goto invalid_attr;
+        if (attr[0]) {
+            if (strcmp(attr[0], "value-format") == 0) {
+                txt_record_value_type value_type;
+
+                if (strcmp(attr[1], "text") == 0) {
```
Comment (pwithnall) [inline avahi-daemon/static-services.c]: I’m not familiar with this XML parsing code, but is it possible for `attr[1]` to be `NULL` here? If so, `strcmp()` will explode.

File: man/avahi.service.5.xml.in [hunk 4]
```diff
@@ -96,7 +96,37 @@
       </option>
 
       <option>
-        <p><opt>&lt;txt-record&gt;</opt> DNS-SD TXT record data.</p>
+        <p><opt>&lt;name replace-wildcards="yes|no"&gt;</opt> The
```
Comment (pwithnall) [inline man/avahi.service.5.xml.in]: This looks unrelated to `txt-record`.

File: man/avahi.service.5.xml.in [hunk 5]
```diff
@@ -96,7 +96,37 @@
...
+        TXT record data. If <opt>value-format</opt> is "text", the
+        value of the TXT record is taken verbatim. If
+        <opt>value-format</opt> is "binary-hex" then the value of TXT
+        record is decoded by taking pairs of characters after the "="
+        char and interpreting them as the textual representation of
+        the two-digit hexadecimal number. Both uppercase and lowercase
+        hexadecimal digits are allowed. The 0x or 0X prefix is not
+        allowed. This requires the length of the value to be even. If
+        <opt>value-format</opt> is "binary-base64" then the value of
+        TXT record is decoded with base64 decoder. The character set
```
Comment (pwithnall) [inline man/avahi.service.5.xml.in]: Nitpick: ‘decoded with _a_ base-64 decoder’.

File: man/avahi.service.5.xml.in [hunk 6]
```diff
@@ -96,7 +96,37 @@
...
+        used is A-Za-z0-9+/. This requires the length of the value to
+        be a multiple of 4, with "=" as a padding. If
```
Comment (pwithnall) [inline man/avahi.service.5.xml.in]: Nitpick: ‘as padding at the end’?

File: avahi-daemon/static-services.c (hunk 2 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: Added.

File: avahi-daemon/static-services.c (hunk 1 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: I forgot to remove that field - it was not used in the end.

File: avahi-daemon/static-services.c (hunk 3 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: Ah, it should not be possible. At even indices you get the attribute name and at odd indices - values. Attributes without values result in parsing error, so we should be safe here.

File: man/avahi.service.5.xml.in (hunk 4 above)
Comment (krnowak) [inline man/avahi.service.5.xml.in]: Hm, I remember copying a paragraph as a help to extend the one about TXT records. Probably forgot to remove it later.

File: man/avahi.service.5.xml.in (hunk 5 above)
Comment (krnowak) [inline man/avahi.service.5.xml.in]: Articles are my bane.

File: man/avahi.service.5.xml.in (hunk 6 above)
Comment (krnowak) [inline man/avahi.service.5.xml.in]: No article before "padding"? ;)

**Comment by krnowak** [comment]
Updated.

File: man/avahi.service.5.xml.in (hunk 6 above)
Comment (pwithnall) [inline man/avahi.service.5.xml.in]: ‘padding’ sounds more natural to me than ‘the padding’. I think it’s because ‘padding’ is uncountable, but your knowledge of English grammar is probably better than mine.

Comment (krnowak) [inline man/avahi.service.5.xml.in]: Alright, I'm going to use the "an" article then.

&lt;/besserwisser&gt;

File: avahi-daemon/static-services.c (hunk 3 above)
Comment (pwithnall) [inline avahi-daemon/static-services.c]: :+1:

File: avahi-daemon/static-services.c [hunk 7]
```diff
@@ -420,6 +449,121 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
...
+
+    if (buf_len % 2) {
+        avahi_log_error("%s: parse failure: hex value of the txt record should have an even length", u->group->filename);
+        u->failed = 1;
+        return -1;
+    }
+    raw_buf_len = buf_len / 2;
+    raw_buf = avahi_malloc(raw_buf_len);
+    for (iter = 0; iter < raw_buf_len; iter++) {
+        uint8_t high_nibble = hex (u->buf[iter * 2]);
```
Comment (pwithnall) [inline avahi-daemon/static-services.c]: Nitpick: There seems to be some inconsistency between `function_name (` and `function_name(` (with and without a space in function calls). You should probably be consistent (I don’t know what the Avahi coding style actually is).

File: avahi-daemon/static-services.c [hunk 8]
```diff
@@ -475,9 +619,46 @@ static void XMLCALL xml_end(void *data, AVAHI_GCC_UNUSED const char *el) {
...
+                        if (decode_base64_buf (u, &value_buf, &value_buf_len) < 0)
+                            return;
+                        free_value_buf = value_buf;
+                        break;
+
+                    default:
+                        assert(0);
+                }
+
+                u->service->txt_records = avahi_string_list_add_anonymous(u->service->txt_records, key_len + 1 + value_buf_len);
```
Comment (pwithnall) [inline avahi-daemon/static-services.c]: Would it make sense to unconditionally nul-terminate the new record? Then any user of Avahi who is just using values with no embedded nuls (for example, ASCII values) can treat them like nul-terminated strings safely.

**Comment by leebyrd** [comment]
+1 on this feature

File: avahi-daemon/static-services.c (hunk 7 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: Yeah, the avahi style is to put no space between function call and the opening parenthesis. It is just my personal style getting the upper hand here. Fixed those.

Comment (krnowak) [inline avahi-daemon/static-services.c]: While replacing that I also spotted one bug, where a free a buffer with `free` instead of `avahi_free`. Fixed that one too.

File: avahi-daemon/static-services.c (hunk 8 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: I don't think so. There is even a function for adding a string (`avahi_string_list_add`) and it only adds `strlen(string)` bytes to the list - so it does not copy the terminating NULL.

**Comment by krnowak** [comment]
Updated.

File: avahi-daemon/static-services.c (hunk 8 above)
Comment (krnowak) [inline avahi-daemon/static-services.c]: I should just add that the `avahi_string_list_add` function is used for the TXT records with the `text` value format (like it was without my changes here), so if somebody was using the record as nul-terminated string, then they were already reading the memory outside of bounds.

Comment (krnowak) [inline avahi-daemon/static-services.c]: Hm, actually, looking at the `avahi_string_list_add_anonymous`, it looks like it allocates a buffer 1 byte larger than requested and terminates the buffer with nul for us.

Comment (pwithnall) [inline avahi-daemon/static-services.c]: Perfect.

**Comment by prestonfick** [comment]
+1 on this feature too. I have a specific use case where we need to specify binary and cannot easily do so in the existing implementation without rewriting and included source.

File: avahi-daemon/static-services.c [hunk 9]
```diff
@@ -420,6 +449,121 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
...
+
+static int base64_error(struct xml_userdata *u, uint8_t *raw_buf) {
+    avahi_log_error("%s: parse failure: failed to parse base64 data: invalid base64 data", u->group->filename);
+    u->failed = 1;
+    avahi_free(raw_buf);
+    return -1;
+}
+
+static int decode_base64_buf(struct xml_userdata *u, uint8_t **out_buf, size_t *out_buf_len) {
+    size_t buf_len = strlen(u->buf);
```
Comment (lathiat) [inline avahi-daemon/static-services.c]: By chance, I tried this record and got a crash here.  Seems it doesn't handle the value being empty which is easy when the base64 value ends in =!

`<txt-record value-format="binary-base64">MDEyMwo=</txt-record>
`

```
(gdb) bt
#0  strlen () at ../sysdeps/x86_64/strlen.S:106
#1  0x0000555555560d6d in decode_base64_buf (u=0x7fffffffe860, out_buf=0x7fffffffe578, out_buf_len=0x7fffffffe580) at static-services.c:516
#2  0x0000555555561378 in xml_end (data=0x7fffffffe860, el=0x5555557874b0 "txt-record") at static-services.c:641
#3  0x00007ffff756fb3c in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#4  0x00007ffff757045c in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#5  0x00007ffff7571d5e in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#6  0x00007ffff7572495 in ?? () from /lib/x86_64-linux-gnu/libexpat.so.1
#7  0x00007ffff757478f in XML_ParseBuffer () from /lib/x86_64-linux-gnu/libexpat.so.1
#8  0x0000555555561c35 in static_service_group_load (g=0x55555577fb50) at static-services.c:828
#9  0x0000555555561d89 in load_file (n=0x55555577fb20 "/services/example.service") at static-services.c:862
#10 0x0000555555561fd4 in static_service_load (in_chroot=1) at static-services.c:919
#11 0x000055555555ceea in run_server (c=0x55555577a200 <config>) at main.c:1235
#12 0x000055555555dc7a in main (argc=1, argv=0x7fffffffeba8) at main.c:1674

```

Comment (krnowak) [inline avahi-daemon/static-services.c]: Actually, `MDEyMwo=` is not base64 encoded value, it is a key `MDEyMwo` that has a base64 encoded empty value. Apparently I forgot to test this case. Thanks for catching this!

**Comment by krnowak** [comment]
Updated the PR to fix the crasher.

File: avahi-daemon/static-services.c [hunk 10]
```diff
@@ -420,6 +449,123 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
     return;
 }
 
+static uint8_t hex(char c)
+{
```
Comment (lathiat) [inline avahi-daemon/static-services.c]: Coding Style, pull { back onto the definition

File: avahi-daemon/static-services.c [hunk 11]
```diff
@@ -420,6 +449,123 @@ static void XMLCALL xml_start(void *data, const char *el, const char *attr[]) {
...
+        }
+        raw_buf[iter] = (high_nibble << 4) | low_nibble;
+    }
+    *out_buf = raw_buf;
+    *out_buf_len = raw_buf_len;
+    return 0;
+}
+
+static uint8_t base64(char c)
+{
```
Comment (lathiat) [inline avahi-daemon/static-services.c]: Coding Style, pull { back onto the definition

**Comment by krnowak** [comment]
Updated to address the coding style issues.

**Comment by lathiat** [comment]
Thanks!
